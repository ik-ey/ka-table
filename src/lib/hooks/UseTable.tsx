import * as actionCreators from '../actionCreators';

import { CustomReducerFunc, OnDispatchFunc } from '../types';
import { ITableInstance, ITableProps } from '../Components/Table/Table';

export const getTable = (options?: {
    changeProps?: React.Dispatch<React.SetStateAction<ITableProps>>;
    onDispatch?: OnDispatchFunc;
    customReducer?: CustomReducerFunc;
}): ITableInstance => {
    const { changeProps, onDispatch, customReducer } = options || {};
    const propsResult = {} as any;
    return {
        ...(Object.keys(actionCreators).reduce((acc, key) => {
            acc[key] = function (...args: any) {
                this.dispatch((actionCreators as any)[key](...args));
            };
            return acc;
        }, {} as any) as typeof actionCreators),
        props: propsResult,
        changeProps: changeProps || (() => {}),
        dispatch: () => {},
        onDispatch: onDispatch || ((() => {}) as OnDispatchFunc),
        customReducer
    };
};

export const useTable = (options?: { onDispatch?: OnDispatchFunc;
                                     customReducer?: CustomReducerFunc; }): ITableInstance => {
    return getTable(options);
};
