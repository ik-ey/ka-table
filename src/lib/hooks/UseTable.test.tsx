import * as React from 'react';

import { Table } from '..';
import { act } from '@testing-library/react';
import { createRoot } from 'react-dom/client';
import { useTable } from './UseTable';

jest.useFakeTimers();

describe('UseTable', () => {
    it('customReducer', () => {
        const customReducer = jest.fn().mockImplementation((state, action) => {
            if (action.type === 'CUSTOM_ACTION'){
                return {...state, data: [{id : 2}]}
            }
        });
        const table = useTable({
            customReducer
        });

        const div = document.createElement('div');
        const root = createRoot(div!);
        act(() => {
            root.render(<Table table={table} rowKeyField='id' columns={[{ key: 'id'}]} data={[]}/>);
        });
        act(() => {
            table.dispatch({ type: 'CUSTOM_ACTION' });
        });
        jest.runAllTimers();
        expect(customReducer).toHaveBeenCalledTimes(2);
        expect(customReducer.mock.calls[1]).toMatchSnapshot();
        expect(customReducer.mock.results[1]).toMatchSnapshot();
    });
});
