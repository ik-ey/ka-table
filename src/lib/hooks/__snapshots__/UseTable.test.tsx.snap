// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UseTable customReducer 1`] = `
Array [
  Object {
    "columns": Array [
      Object {
        "key": "id",
      },
    ],
    "data": Array [],
    "rowKeyField": "id",
  },
  Object {
    "type": "CUSTOM_ACTION",
  },
  Object {
    "columns": Array [
      Object {
        "key": "id",
      },
    ],
    "data": Array [],
    "rowKeyField": "id",
  },
]
`;

exports[`UseTable customReducer 2`] = `
Object {
  "type": "return",
  "value": Object {
    "columns": Array [
      Object {
        "key": "id",
      },
    ],
    "data": Array [
      Object {
        "id": 2,
      },
    ],
    "rowKeyField": "id",
  },
}
`;
